reconcile/openshift_prometheusrules.py

```diff
@@ -0,0 +1,273 @@
+
+    # percentile is mandatory for latency rates
+    for rule in [r for r in pp['sloRules'] if r['kind'] == 'latency_rate']:
+        if 'percentile' not in rule:
+            errors.append('percentile missing in %s slo rule' % rule['name'])
+
+    # volume, latency, errors => check that rules exist in slo_recordings
+    # we'll also use it for the availability validation
+    slos = {}
+    for category in ['volume', 'latency', 'errors']:
+        slos[category] = set([s['name'] for s in pp[category]])
+
+        if len(slos[category]) != len(pp[category]):
+            errors.append('slo names are not unique for %s' % category)
+
+        for slx in pp[category]:
+            if slx['rule'] not in slo_rules:
+                errors.append('Unknown slo rule %s' % slx['rule'])
+
+    # check availability names are unique and slo rules exist
+    availability_rule_names = set()
+    for slx in pp['availability']:
+        availability_rule_names.add(slx['name'])
+        for c in ['latency', 'errors']:
+            for rule_name in slx['rules'][category]:
+                if rule_name not in slos[category]:
+                    errors.append('Unknown %s rule %s' % (category,
+                                                          rule_name))
+
+    if len(availability_rule_names) != len(pp[category]):
+        errors.append('slo names are not unique for %s' % category)
+
+    return errors
+
+
+# Build params to pass to the template
+def build_template_params(pp):
+    params = {}
+    params['http_rates'] = []     # sloRules of http_rate type
+    params['latency_rates'] = []  # sloRules of latency_rate type
+    params['all_rules'] = []      # contains the name of each rule.
+
+    # We have checked that rule names are unique by category, but not as whole
+    # we will add a suffix to each rule to make sure they're unique
+    for r in pp['sloRules']:
+        if r['kind'] == 'http_rate':
+            params['http_rates'].append(r)
+            params['all_rules'].extend(
+                ['%s__http_rates.rateRules' % r['name'],
+                 '%s__http_rates.errorRateRules' % r['name']])
+        else:
+            params['latency_rates'].append(r)
+            params['all_rules'].append('%s__latency_rates.rules' % r['name'])
+
+    for c in ['volume', 'latency', 'errors', 'availability']:
+        params['all_rules'].extend(['%s__%s.rules' % (r['name'], c)
+                                    for r in pp[c]])
+        params[c] = pp[c]
+
+    params['labels'] = pp['labels']
+    params['component'] = pp['component']
+    params['namespace'] = pp['namespace']['name']
+    params['prometheus_labels'] = pp['prometheusLabels']
+
+    return params
+
+
+def fetch_desired_state(performance_parameters, ri):
+    for pp in performance_parameters['performance_parameters_v1']:
+        if pp['namespace']['cluster']['observabilityNamespace'] is None:
+            logging.error('No observability namespace for %s' % pp['name'])
```

Comment:
we are logging many errors, but the integration will still exit with a 0 exit code.
you can use `ri.register_error()` and check `ri.has_error_registered()` in the run function.
if True, let's still act on everything we can, but exit with an exit code so we are aware there is a problem.

Reply:
+1

Reply:
Hmmm, I'm trying to avoid that an error in one performance parameter does not affect the rest. So if one fails and the rest goes well I should be even returning an exit code > 0? I'm inclined to yes.

The main problem is that some of the validations here should be done in app-interface. I guess this would help :) https://issues.redhat.com/browse/APPSRE-315

Reply:
ahh, @jmelis has reminded me that we run the integrations at PR time... so I guess we're better covered :-D

Reply:
fixed in 6bbfd5a